src/frontend/specification/ParameterTable.jl

```diff
@@ -124,128 +153,182 @@ end
 
 Base.showerror(io::IO, e::CyclicModelError) = print(io, e.msg)
 
-import Base.sort!, Base.sort
+"""
+    sort_vars!(partable::ParameterTable)
+    sort_vars!(partables::EnsembleParameterTable)
 
-function sort!(partable::ParameterTable)
-    variables = [partable.variables[:latent_vars]; partable.variables[:observed_vars]]
+Sort variables in `partable` so that all independent variables are
+before the dependent variables and store it in `partable.variables.sorted`.
 
-    is_regression =
-        (partable.columns[:parameter_type] .== :→) .&
-        (partable.columns[:from] .!= Symbol("1"))
+If the relations between the variables are acyclic, sorting will
+make the resulting `A` matrix in the *RAM* model lower triangular
+and allow faster calculations.
+"""
+function sort_vars!(partable::ParameterTable)
+    vars = [
+        partable.latent_vars
+        partable.observed_vars
+    ]
 
-    to = partable.columns[:to][is_regression]
-    from = partable.columns[:from][is_regression]
+    # regression edges (excluding intercept)
+    edges = [
+        (from, to) for (rel, from, to) in zip(
+            partable.columns[:relation],
+            partable.columns[:from],
+            partable.columns[:to],
+        ) if (rel == :→) && (from != Symbol("1"))
+    ]
+    sort!(edges, by = last) # sort edges by target
 
-    sorted_variables = Vector{Symbol}()
+    sorted_vars = Vector{Symbol}()
 
-    sorted = false
-    while !sorted
+    while !isempty(vars)
         acyclic = false
 
-        for (i, variable) in enumerate(variables)
-            if !(variable ∈ to)
-                push!(sorted_variables, variable)
-                deleteat!(variables, i)
-                delete_edges = from .!= variable
-                to = to[delete_edges]
-                from = from[delete_edges]
+        for (i, var) in enumerate(vars)
+            # check if var has any incoming edge
+            eix = searchsortedfirst(edges, (var, var), by = last)
+            if !(eix <= length(edges) && last(edges[eix]) == var)
```

Comment:
I think you can use `findfirst` here to avoid the complicated checks in the if-condition
```suggestion
            eix = findfirst(e -> e[2] == var, edges)
            if isnothing(eix)
```

Reply:
The point of using `searchsortedfirst()` is to find an edge with *O(log N)* comparisons (the edges are sorted by the incoming vertex) instead of *O(N)* that you get with `findfirst()`.